nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java

```diff
@@ -0,0 +1,232 @@
+            .name("delimiter")
+            .displayName("Delimiter")
+            .expressionLanguageSupported(false)
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .description("The string used to delimit directories within the bucket. Please consult the AWS documentation " +
+                    "for the correct use of this field.")
+            .build();
+
+    public static final PropertyDescriptor PREFIX = new PropertyDescriptor.Builder()
+            .name("prefix")
+            .displayName("Prefix")
+            .expressionLanguageSupported(true)
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .description("The prefix used to filter the object list. In most cases, it should end with a forward slash ('/').")
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(BUCKET, KEY, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE, TIMEOUT, SSL_CONTEXT_SERVICE, ENDPOINT_OVERRIDE,
+                    PROXY_HOST, PROXY_HOST_PORT, DELIMITER, PREFIX));
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(
+            new HashSet<>(Collections.singletonList(REL_SUCCESS)));
+
+    public static final String CURRENT_TIMESTAMP = "currentTimestamp";
+    public static final String CURRENT_KEY_PREFIX = "key-";
+
+    // State tracking
+    private long currentTimestamp = 0L;
+    private Set<String> currentKeys;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    private Set<String> extractKeys(final StateMap stateMap) {
+        Set<String> keys = new HashSet<>();
+        for (Map.Entry<String, String>  entry : stateMap.toMap().entrySet()) {
+            if (entry.getKey().startsWith(CURRENT_KEY_PREFIX)) {
+                keys.add(entry.getValue());
+            }
+        }
+        return keys;
+    }
+
+    private void restoreState(final ProcessContext context) throws IOException {
+        final StateMap stateMap = context.getStateManager().getState(Scope.CLUSTER);
+        if (stateMap.getVersion() == -1L || stateMap.get(CURRENT_TIMESTAMP) == null || stateMap.get(CURRENT_KEY_PREFIX) == null) {
+            currentTimestamp = 0L;
+            currentKeys = new HashSet<>();
+        } else {
+            currentTimestamp = Long.parseLong(stateMap.get(CURRENT_TIMESTAMP));
+            currentKeys = extractKeys(stateMap);
+        }
+    }
+
+    private void persistState(final ProcessContext context) {
+        Map<String, String> state = new HashMap<>();
+        state.put(CURRENT_TIMESTAMP, String.valueOf(currentTimestamp));
+        int i = 0;
```

Comment:
This seems like it might be a bit off as "i" never gets incremented and the state would just be overwriting for each of the keys currently in there.  What is the intended approach?